boot/freeldr/freeldr/ntldr/winldr.c

```diff
@@ -629,65 +689,24 @@ LoadWindowsCore(IN USHORT OperatingSystemVersion,
         SosEnabled = TRUE;
     }
 
-    if (OperatingSystemVersion > _WIN32_WINNT_NT4)
+    if (OperatingSystemVersion > _WIN32_WINNT_NT4 &&
+        NtLdrGetOption(BootOptions, "SAFEBOOT"))
     {
-        if (NtLdrGetOption(BootOptions, "SAFEBOOT"))
-        {
-            /* We found the SAFEBOOT option. */
-            FIXME("LoadWindowsCore: SAFEBOOT - TRUE (not implemented)\n");
-            SafeBoot = TRUE;
-        }
-        if (NtLdrGetOption(BootOptions, "PAE"))
-        {
-            /* We found the PAE option. */
-            FIXME("LoadWindowsCore: PAE - TRUE (not implemented)\n");
-            PaeEnabled = TRUE;
-        }
+        /* We found the SAFEBOOT option. */
+        FIXME("LoadWindowsCore: SAFEBOOT - TRUE (not implemented)\n");
+        SafeBoot = TRUE;
     }
 
-    if (OperatingSystemVersion > _WIN32_WINNT_WIN2K)
+    if (OperatingSystemVersion > _WIN32_WINNT_WIN2K &&
+        NtLdrGetOption(BootOptions, "BOOTLOGO"))
     {
-        if (NtLdrGetOption(BootOptions, "NOPAE"))
-        {
-            /* We found the NOPAE option. */
-            FIXME("LoadWindowsCore: NOPAE - TRUE (not implemented)\n");
-            PaeDisabled = TRUE;
-        }
-        if (NtLdrGetOption(BootOptions, "BOOTLOGO"))
-        {
-            /* We found the BOOTLOGO option. */
-            FIXME("LoadWindowsCore: BOOTLOGO - TRUE (not implemented)\n");
-            BootLogo = TRUE;
-        }
-
-        if (!LoaderBlock->SetupLdrBlock)
-        {
-            if (NtLdrGetOption(BootOptions, "EXECUTE"))
-            {
-                /* We found the EXECUTE option. */
-                FIXME("LoadWindowsCore: EXECUTE - TRUE (not implemented)\n");
-                NoexecuteDisabled = TRUE;
-            }
-            if (NtLdrGetOption(BootOptions, "NOEXECUTE=ALWAYSOFF"))
-            {
-                /* We found the NOEXECUTE=ALWAYSOFF option. */
-                FIXME("LoadWindowsCore: NOEXECUTE=ALWAYSOFF - TRUE (not implemented)\n");
-                NoexecuteDisabled = TRUE;
-            }
-            if (NtLdrGetOption(BootOptions, "NOEXECUTE"))
-            {
-                /* We found the NOEXECUTE option. */
-                FIXME("LoadWindowsCore: NOEXECUTE - TRUE (not implemented)\n");
-                NoexecuteEnabled = TRUE;
-            }
-        }
+        /* We found the BOOTLOGO option. */
+        FIXME("LoadWindowsCore: BOOTLOGO - TRUE (not implemented)\n");
+        BootLogo = TRUE;
     }
 
-    if (SafeBoot)
-    {
-        PaeDisabled = TRUE;
-        NoexecuteDisabled = TRUE;
-    }
+    PaeModeOn = WinLdrIsPaeSupported(LoaderBlock, BootOptions, OperatingSystemVersion, KernelFileName, HalFileName);
```

Comment:
It would be actually better to move this `WinLdrIsPaeSupported` call at the very beginning of the `LoadWindowsCore` function, so that we could then, depending on whether PAE is enabled or not, to load the appropriate kernel file.

Reply:
Quite right. That I me meant when I offered to divide everything you need for "PAE for Freeloader" for three parts.
Namely:
The 1st part is the "physics" paradise, that is, directly functional things. It is already ready.
The 2nd part is the "logic" PAE. These are different checks, as a result of which the decision is made to what mode (PAE or Non-PAE) is included. It also includes checking and selecting the kernel file for the appropriate mode. I can not agree that it's easy. Perhaps this is the most costly part.
The 3rd part is the refactoring code and the like (there are no functional changes). I consider this part of the trivial and not having a high priority.